### ISSUE TITLE
Ability to block until deploy and port-forward are finished

### ISSUE THREAD

**Issue by cpoole** [GoogleContainerTools/skaffold#4022]
### Expected behavior
For usage in CI pipelines (specifically integration testing) it would be nice if there was an easy way to do `skaffold run --port-forward` and _not_ have it hang indefinitely. Instead it would simply block until all port forwards have been completed. 

Currently I run `skaffold run` then manually create the port forwards, but now i've got these port forwards duplicated.

Is this a feature anyone else is interested in? Is the some easier, already existing way to accomplish this?

**Comment by tstromberg** [comment]
What do you mean by `block until port forwards have been completed`?

The tricky part, AFAIK, is that the port-forward process has to keep running. Are you suggesting that it daemonizes them?

**Comment by briandealwis** [comment]
You can use the events API to be notified of established port-forwards.

We'd like to hear more about how you're doing testing, however!  This is an area we'd like to improve in Skaffold.

**Comment by cpoole** [comment]
Yes, I think demonizing them would be ideal. Let me explain my current workflow:

- git trigger to jenkins + spin up new node
- bootstrap new node by installing kind 
- skaffold run to install my app in kind
- little dumb bash script to find all of the services spun up by skaffold and fork a `kubectl port-forward` for each of them (append PID's to a file). 
- run my integration tests against the stack in kind
- kill all the pid's 
- kill kind

Pretty standard test cycle imo. What I'd like to see is some sort of option for skaffold to daemonize those port forwards and then know how to properly kill them with some other command `skaffold delete --port-forwards` or something like that. 

Effectively mimic the behavior of running `docker-compose up -d` and then `docker-compose down` after tests.

**Comment by cpoole** [comment]
Listening to the events is probably fine @briandealwis . I imagine id just fork a skaffold dev and block while listening. Is there an event for "all forwards done"?

Another way to solve this would be sort of `skaffold test` pipeline where we could pass in an arbitrary set of commands for skaffold to run after finishing a standard skaffold dev. But I've seen resistance (understandably) into making skaffold a generic workflow engine.

**Comment by tejal29** [comment]
Currently we don't have an event for Portforward completed. 
We do have a an event for `DevLoopEvent ` complete which can indicate when an entire devloop is complete which is sent after following phases are complete
1. Build,
2. Deploy, 
3. Status check - if `--status-check` is true
4. port forwarding complete - if `--port-forward` is true.

Can you try that ?

**Comment by nkubala** [comment]
@cpoole FWIW `skaffold test` is still in consideration as a potential phase, and I think you'd be able to accomplish what you need here with it.

cc #992

**Comment by gnarea** [comment]
I think the feature proposed here offers a better way to run a functional test suite in CI, and I don't see myself using `skaffold test` for the reasons here: https://github.com/GoogleContainerTools/skaffold/issues/992#issuecomment-729993870

As a workaround, I'm having to run `skaffold run` and then manually call `kubectl port-forward` for each port I need to access from the tests: https://github.com/relaycorp/relaynet-internet-gateway/blob/2eae4449267a5ecdab85c62ae2c0bfe27458cdd0/.github/workflows/ci-cd.yml#L42-L66

**Comment by tejal29** [comment]
We have plans to improve `skaffold test` later in the year. 
this could also been done via hooks #1441 that @gsquared94 is implementing.

@gsquared94 do you have an idea if post deploy hooks will run after port-forwarfing is stable ?

**Comment by ethanmdavidson** [comment]
The post-deploy hook appears to run before the port-forwarding is stable, unfortunately.
